### ISSUE TITLE
How to increase max streams (websockets) limit?

### ISSUE THREAD

**Issue by FelipeEmerim** [envoyproxy/envoy#20020]
*Title*: *How to increase max streams (websockets) limit?*

*Description*:
We are having a problem using consul + envoy sidecar with websockets. If we have about +1024-1100 open websocket connections at the same time, envoy starts dropping new incoming connections. Because of that our healthchecks (kafka ping) start failing and the service becomes unreachable. If we remove the envoy sidecar from the pods it starts accepting as many connections as it can.

We think envoy is blocking the connections because of the following logs:
```
[2022-02-16 19:42:27.155][15][debug][pool] [source/common/conn_pool/conn_pool_base.cc:239] [C5221] using existing connection
[2022-02-16 19:42:27.155][15][debug][pool] [source/common/conn_pool/conn_pool_base.cc:170] max streams overflow
[2022-02-16 19:42:27.155][15][debug][router] [source/common/router/router.cc:1130] [C5228][S6567386237767641786] upstream reset: reset reason: overflow, transport failure reason: 
[2022-02-16 19:42:27.155][15][debug][http] [source/common/http/filter_manager.cc:947] [C5228][S6567386237767641786] Sending local reply with details upstream_reset_before_response_started{overflow}
[2022-02-16 19:42:27.155][15][debug][http] [source/common/http/conn_manager_impl.cc:1412] [C5228][S6567386237767641786] closing connection due to connection close header
[2022-02-16 19:42:27.155][15][debug][http] [source/common/http/conn_manager_impl.cc:1467] [C5228][S6567386237767641786] encoding headers via codec (end_stream=false):
```

Is there a way to increase the concurrent stream limit on envoy? Could we make it so that k8s probes connections are not affected by this behavior?

[optional *Relevant Links*:]
We also opened an issue in consul to help investigate the issue: https://github.com/hashicorp/consul/issues/12373. We opened this issue to understand why envoy behaves this way and how to change it, so that we can find out which parameter we need to change in consul.

**Comment by KBaichoo** [comment]
are you hitting the [circuit breakers](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/circuit_breaking.html?highlight=circuit%20breaker)? they are defaulted to 1024 connections per cluster and enabled by default. You can configure them to be higher

**Comment by FelipeEmerim** [comment]
We thought we could be hitting **Cluster maximum connections** but consul is supposed to expose that throught the service defaults CRD https://www.consul.io/docs/connect/config-entries/service-defaults . We changed `UpstreamConfig.defaults.limits`, increasing all values to 10 million. This change affected only HTTP connections though, TCP streams remained limited in about 1024.

According to these links:

* https://learn.hashicorp.com/tutorials/consul/service-mesh-circuit-breaking?in=consul/developer-mesh
* https://github.com/hashicorp/consul/issues/7211

Service defaults should configure the circuit breaker.

Do circuit breakers also apply to TCP traffic and streams? If so then consul is doing something wrong on their end.

**Comment by KBaichoo** [comment]
is it possible for you to explictly set circuit breakers to effectively be disabled ala: https://www.envoyproxy.io/docs/envoy/latest/faq/load_balancing/disable_circuit_breaking#faq-disable-circuit-breaking ?

Circuit breakers apply to clusters. Perhaps check if `upstream_cx_overflow` stat is incremented on your cluster?

**Comment by FelipeEmerim** [comment]
I will try that on Monday and post the results here. Is there an endpoint on envoy where I can see the value of the config you mentioned?

**Comment by KBaichoo** [comment]
yes, the stats endpoint from the admin: https://www.envoyproxy.io/docs/envoy/latest/operations/admin#get--stats, there's also a config dump which might be useful to see the settings the envoy has: https://www.envoyproxy.io/docs/envoy/latest/operations/admin#get--config_dump

**Comment by FelipeEmerim** [comment]
I performed some tests using the endpoints you mentioned. It seems Consul is updating the circuit breakers on most of the clusters. I can see every other pod that has the sidecar injected in the `dynamic_active_clusters` key.

When I apply the service defaults, each item gets the following circuit breaker config:

```json
"circuit_breakers": {
  "thresholds": [
     {
       "max_connections": 10000000,
       "max_pending_requests": 10000000,
       "max_requests": 10000000
     }
   ]
},
```

I applied this on both my service and the ingress controller, and I still hit the limit mentioned in the question earlier.  The readiness probe stops working (it returns 503 status code) and the service goes down.

I also checked the stats endpoint you mentioned and found only one cluster with upstream_cx_overflow greater than 0.

```json
{
    "name": "cluster.local_app.upstream_cx_overflow",
    "value": 17
},
```

We do have this `local_app` in our config:


```json
"version_info": "25f00611228c29a0e12ab6a09c01822b735f910d0aedd4d8cd6ed1b6f28f99e8",
"cluster": {
    "@type": "type.googleapis.com/envoy.config.cluster.v3.Cluster",
    "name": "local_app",
    "type": "STATIC",
    "connect_timeout": "5s",
    "load_assignment": {
        "cluster_name": "local_app",
        "endpoints": [
            {
                "lb_endpoints": [
                    {
                        "endpoint": {
                            "address": {
                                "socket_address": {
                                    "address": "127.0.0.1",
                                    "port_value": 3000
                                }
                            }
                        }
                    }
                ]
            }
        ]
    }
},
"last_updated": "2022-02-22T16:34:57.293Z"
},
```
I see no circuit breakers in the local_app cluster, could it be that this is causing our issues?

**Comment by KBaichoo** [comment]
```
{
    "name": "cluster.local_app.upstream_cx_overflow",
    "value": 17
},
```
leads me to think we are hitting the circuit break default connection limits of 1024 for this cluster, and the config for the `local_app` cluster doesn't seem to config the circuit breakers themselves, so we're probably just using the default of 1024.

Can you try configuring the circuit breaker for `local_app`?

**Comment by FelipeEmerim** [comment]
Unfortunately Consul does not allow me to change that cluster because of a limitation in their config override. Every pod has a `local_app` cluster and I cannot supply the `port_value` dynamically in the config override. The override is global so we would need to have every pod using the same service port.

That being said, I believe we can close this question since we have got concrete evidence of what is causing the problem. I opened a follow-up issue on Consul linking this discussion.

Thank you so much for all the help, thanks to you we could find the root cause of the problem and learn a little bit about envoy.
